cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,313 @@
+        }
+    };
+}
+
+void UUrdfRobotComponent::TickComponent(float DeltaTime, enum ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+    USceneComponent::TickComponent(DeltaTime, TickType, ThisTickFunction);
+
+    if (!ticked_once_) {
+        SP_LOG_CURRENT_FUNCTION();
+        initializeDeferred();
+        ticked_once_ = true;
+    }
+
+    // Update this component's pose to match root link component
+    SP_ASSERT(root_link_component_);
+    bool sweep = false;
+    FHitResult* hit_result = nullptr;
+    SetWorldLocationAndRotation(root_link_component_->GetComponentLocation(), root_link_component_->GetComponentRotation(), sweep, hit_result, ETeleportType::None);
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+
+    if (Std::contains(action_components_, "control_joints")) {
+        for (auto joint_component : joint_components_) {
+            action_space.merge(joint_component.second->getActionSpace());
+        }
+    }
+
+    return action_space;
+}
+
+std::map<std::string, ArrayDesc> UUrdfRobotComponent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
+
+    if (Std::contains(observation_components_, "link_configurations")) {
+        for (auto link_component : LinkComponents) {
+            observation_space.merge(link_component->getObservationSpace());
+        }
+    }
+
+    if (Std::contains(observation_components_, "joint_configurations")) {
+        for (auto joint_component : JointComponents) {
+            observation_space.merge(joint_component->getObservationSpace());
+        }
+    }
+
+    return observation_space;
+}
+
+void UUrdfRobotComponent::applyAction(const std::map<std::string, std::vector<uint8_t>>& actions)
+{
+    std::map <std::string, std::vector<double>> joint_actions;
+
+    for (auto& action : actions) {
+        std::vector<double> action_data = Std::reinterpretAs<double>(action.second);
+        joint_actions[action.first] = action_data;
+    }
+
+    bool assert_if_action_is_inconsistent_with_joint = true;
+    bool assert_if_joint_not_found = true;
+    applyAction(joint_actions, assert_if_joint_not_found, assert_if_action_is_inconsistent_with_joint);
+}
+
+std::map<std::string, std::vector<uint8_t>> UUrdfRobotComponent::getObservation() const
+{
+    std::map<std::string, std::vector<uint8_t>> observation;
+
+    if (Std::contains(observation_components_, "link_configurations")) {
+        for (auto link_component : LinkComponents) {
+            observation.merge(link_component->getObservation());
+        }
+    }
+
+    if (Std::contains(observation_components_, "joint_configurations")) {
+        for (auto joint_component : JointComponents) {
+            observation.merge(joint_component->getObservation());
+        }
+    }
+
+    return observation;
+}
+
+void UUrdfRobotComponent::reset()
+{
+    for (auto link_component : LinkComponents) {
```

Comment:
We should adopt a consistent policy in this file for all of these for loops. I don't care if we iterate over `LinkComponents` or `link_components_`, but we should do it consistently.

```
    for (auto& link_component : link_components_) {
        link_components_.second->reset();
    }
```